File: server/controllers/newsletter_controller.py
from fastapi import HTTPException
from datetime import datetime
from pymongo import MongoClient
import os

# Get MongoDB connection
MONGO_URI = os.getenv("MONGO_URI", "mongodb://localhost:27017")
client = MongoClient(MONGO_URI)
db = client["ecommerce"]

# Subscribe to newsletter
async def subscribe_newsletter(data: dict, user=None):
    try:
        email = data.get("email", "").strip().lower()
        if not email:
            raise HTTPException(status_code=400, detail="Email is required")

        coll = db.get_collection("newsletter_subscriptions")
        
        # Check if already subscribed
        existing = coll.find_one({"email": email})
        
        if existing:
            if existing.get("status") == "unsubscribed":
                # Resubscribe
                coll.update_one(
                    {"email": email},
                    {"$set": {"status": "subscribed", "unsubscribed_at": None, "updated_at": datetime.utcnow()}}
                )
                return {
                    "success": True,
                    "message": "Welcome back! You have been resubscribed to our newsletter"
                }
            return {
                "success": True,
                "message": "Email already subscribed"
            }

        # Create new subscription
        subscription = {
            "email": email,
            "status": "subscribed",
            "verified": True,
            "verified_at": datetime.utcnow(),
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        }

        result = coll.insert_one(subscription)
        return {
            "success": True,
            "message": "You have been subscribed to our newsletter",
            "id": str(result.inserted_id)
        }
    except HTTPException:
        raise
    except Exception as e:
        print(f"Error subscribing to newsletter: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error subscribing to newsletter: {str(e)}")

# Get all newsletter subscribers
async def get_subscribers(user=None):
    try:
        from bson import ObjectId
        # Helper to serialize documents
        def _serialize(doc):
            if doc is None:
                return None
            def _walk(o):
                if isinstance(o, ObjectId):
                    return str(o)
                if isinstance(o, datetime):
                    return o.isoformat()
                if isinstance(o, dict):
                    return {k: _walk(v) for k, v in o.items()}
                if isinstance(o, list):
                    return [_walk(i) for i in o]
                return o
            return _walk(doc)
        
        coll = db.get_collection("newsletter_subscriptions")
        cursor = coll.find({}).sort("_id", -1)
        subscribers = [_serialize(doc) for doc in cursor]
        return subscribers
    except Exception as e:
        print(f"Error fetching subscribers: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching subscribers: {str(e)}")

# Delete a subscriber
async def delete_subscriber(subscriber_id: str, user=None):
    try:
        from bson import ObjectId
        coll = db.get_collection("newsletter_subscriptions")
        sub_oid = ObjectId(subscriber_id) if ObjectId.is_valid(subscriber_id) else None
        if not sub_oid:
            raise HTTPException(status_code=400, detail="Invalid subscriber ID")
        
        result = coll.find_one_and_delete({"_id": sub_oid})
        if not result:
            raise HTTPException(status_code=404, detail="Subscriber not found")
        
        return {"message": "Subscriber deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        print(f"Error deleting subscriber: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error deleting subscriber: {str(e)}")

File: server/controllers/newsletterController.js
const NewsletterSubscription = require('../models/NewsletterSubscription');

/**
 * Subscribe to newsletter
 */
const subscribeNewsletter = async (req, res) => {
  try {
    const { email } = req.body;

    if (!email) {
      return res.status(400).json({
        success: false,
        message: 'Email is required'
      });
    }

    // Check if already subscribed
    let subscription = await NewsletterSubscription.findOne({ email: email.toLowerCase() });

    if (subscription) {
      // If unsubscribed, resubscribe
      if (subscription.status === 'unsubscribed') {
        subscription.status = 'subscribed';
        subscription.unsubscribedAt = null;
        await subscription.save();
        return res.status(200).json({
          success: true,
          message: 'Welcome back! You have been resubscribed to our newsletter',
          data: subscription
        });
      }
      // If already subscribed
      return res.status(200).json({
        success: true,
        message: 'Email already subscribed',
        data: subscription
      });
    }

    // Create new subscription
    subscription = new NewsletterSubscription({
      email: email.toLowerCase(),
      status: 'subscribed',
      verified: true, // Auto-verify for now (can implement email verification later)
      verifiedAt: new Date(),
      ipAddress: req.ip,
      userAgent: req.get('user-agent')
    });

    await subscription.save();

    res.status(201).json({
      success: true,
      message: 'Successfully subscribed to newsletter',
      data: subscription
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error subscribing to newsletter'
    });
  }
};

/**
 * Get all newsletter subscriptions (Admin only)
 */
const getNewsletterSubscriptions = async (req, res) => {
  try {
    const { status = 'subscribed', limit = 50, skip = 0 } = req.query;

    const query = {};
    if (status) {
      query.status = status;
    }

    const subscriptions = await NewsletterSubscription.find(query)
      .sort({ createdAt: -1 })
      .skip(parseInt(skip))
      .limit(parseInt(limit));

    const total = await NewsletterSubscription.countDocuments(query);

    res.status(200).json({
      success: true,
      data: subscriptions,
      total,
      count: subscriptions.length
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching newsletter subscriptions'
    });
  }
};

/**
 * Get single newsletter subscription
 */
const getNewsletterSubscription = async (req, res) => {
  try {
    const { id } = req.params;

    const subscription = await NewsletterSubscription.findById(id);

    if (!subscription) {
      return res.status(404).json({
        success: false,
        message: 'Subscription not found'
      });
    }

    res.status(200).json({
      success: true,
      data: subscription
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching subscription'
    });
  }
};

/**
 * Unsubscribe from newsletter
 */
const unsubscribeNewsletter = async (req, res) => {
  try {
    const { email } = req.body;

    if (!email) {
      return res.status(400).json({
        success: false,
        message: 'Email is required'
      });
    }

    const subscription = await NewsletterSubscription.findOneAndUpdate(
      { email: email.toLowerCase() },
      {
        status: 'unsubscribed',
        unsubscribedAt: new Date()
      },
      { new: true }
    );

    if (!subscription) {
      return res.status(404).json({
        success: false,
        message: 'Email not found in newsletter'
      });
    }

    res.status(200).json({
      success: true,
      message: 'Successfully unsubscribed from newsletter',
      data: subscription
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error unsubscribing from newsletter'
    });
  }
};

/**
 * Update newsletter subscription status (Admin only)
 */
const updateNewsletterStatus = async (req, res) => {
  try {
    const { id } = req.params;
    const { status } = req.body;

    if (!status || !['subscribed', 'unsubscribed', 'bounced'].includes(status)) {
      return res.status(400).json({
        success: false,
        message: 'Valid status required (subscribed, unsubscribed, bounced)'
      });
    }

    const subscription = await NewsletterSubscription.findByIdAndUpdate(
      id,
      {
        status,
        unsubscribedAt: status === 'unsubscribed' ? new Date() : null
      },
      { new: true }
    );

    if (!subscription) {
      return res.status(404).json({
        success: false,
        message: 'Subscription not found'
      });
    }

    res.status(200).json({
      success: true,
      message: 'Subscription status updated',
      data: subscription
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error updating subscription'
    });
  }
};

/**
 * Delete newsletter subscription (Admin only)
 */
const deleteNewsletterSubscription = async (req, res) => {
  try {
    const { id } = req.params;

    const subscription = await NewsletterSubscription.findByIdAndDelete(id);

    if (!subscription) {
      return res.status(404).json({
        success: false,
        message: 'Subscription not found'
      });
    }

    res.status(200).json({
      success: true,
      message: 'Subscription deleted'
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error deleting subscription'
    });
  }
};

/**
 * Get newsletter statistics (Admin only)
 */
const getNewsletterStats = async (req, res) => {
  try {
    const total = await NewsletterSubscription.countDocuments();
    const subscribed = await NewsletterSubscription.countDocuments({ status: 'subscribed' });
    const unsubscribed = await NewsletterSubscription.countDocuments({ status: 'unsubscribed' });
    const bounced = await NewsletterSubscription.countDocuments({ status: 'bounced' });

    res.status(200).json({
      success: true,
      data: {
        total,
        subscribed,
        unsubscribed,
        bounced,
        subscriptionRate: total > 0 ? ((subscribed / total) * 100).toFixed(2) : 0
      }
    });
  } catch (error) {

    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching statistics'
    });
  }
};

module.exports = {
  subscribeNewsletter,
  getNewsletterSubscriptions,
  getNewsletterSubscription,
  unsubscribeNewsletter,
  updateNewsletterStatus,
  deleteNewsletterSubscription,
  getNewsletterStats
};

